### PR TITLE
#11514 - Svg Plan throws no instance of an object error on drawing pa…

### PR DESCRIPTION
**Pull request by Zeljko-Predjeskovic**
The issue is that the line segments after close segment Z have a start path of null but it should start from the first move M.

If the Start path is null, we set it to the first path from graphicspath. 

The first path is always a move M no other path segment can stay fist.
grafik

### CODE REVIEW

**Review by gentledepp — APPROVED** [review]
Looks good!

**Comment by Zeljko-Predjeskovic** [comment]
Not merging yet... found a new svg bug :D

**Comment by Zeljko-Predjeskovic** [comment]
image

Either we get the bound of the clip, or the elements bounds if they are smaller than the clippaths bounds.

When the element bounds are much much bigger than the clips, we still pick the clips bounds. It cannot go outside of the clip:

image

File: Svg.Tests.Win/SvgClipPathTest.cs [hunk 1]
```diff
@@ -68,4 +68,18 @@ public void WhenGroupWithClipAndChildWithMatrixRenderer_CanvasIsSavedCorrectly()
         c.AssertAreSimilar(97f, svgPath);
 
     }
+
+    [Test]
+    public void CreatingAClip_WithARectangle_GetClipBounds()
```
Comment (gentledepp) [inline Svg.Tests.Win/SvgClipPathTest.cs]: please come up with a better name.
Tests should (please) be named like:
`<operation/action>_<precondition>_<expectedoutcome>`

So like
`WhenCalculatingBounds_AndRectangleIsClippedWhereClipIsSmallerThanTheRectangle_ReturnsTheClipBounds`

File: Svg.Tests.Win/SvgDocumentTest.cs [hunk 2]
```diff
@@ -156,5 +156,17 @@ public void WhenDocumentOpenWithCss_ElementsAdaptStyle()
             actualFontFamily[0].Should().Be("sans-serif");
             actualFontFamily[1].Should().Be("");
         }
+
+        [Test]
+        public void WhenPathHasClip_CombineBoundsWithPathAndClip()
```
Comment (gentledepp) [inline Svg.Tests.Win/SvgDocumentTest.cs]: again: I cannot understand what is being tested here.
That a clipped path's bounds are calculated by _merging_ the bounds of the clip and the path?

Remember:
`<operation/action>_<precondition>_<expectedoutcome>`

File: Svg.Tests.Win/SvgVisualElementsExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+public class SvgVisualElementsExtensionsTests
+{
+    [SetUp]
+    public void SetUp()
+    {
+        SvgPlatform.Init();
+    }
+
+    [Test]
+    public void WhenGettingBoundsFromVisualElement_WithClipPath_GetClipPathBound()
```
Comment (gentledepp) [inline Svg.Tests.Win/SvgVisualElementsExtensionsTests.cs]: Better naming,
But this implies that _alway_ the clippath bounds should be used.
Is that the case?

File: Svg.Tests.Win/SvgVisualElementsExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        Assert.IsNotNull(rectangle);
+        Assert.AreEqual(20000, (int)rectangle.Width);
+        Assert.AreEqual(100000, (int)rectangle.Height);
+        Assert.AreEqual(200, rectangle.GetBounds().Width);
+        Assert.AreEqual(50, rectangle.GetBounds().Height);
+        
+    }
+
+    [Test]
+    public void WhenGettingBoundsFromVisualElement_WithClipPathBiggerThanElement_GetElementBound()
```
Comment (gentledepp) [inline Svg.Tests.Win/SvgVisualElementsExtensionsTests.cs:35]: this is a great name!! ♥️

File: Svg/Basic Shapes/SvgImage.cs [hunk 5]
```diff
@@ -97,26 +97,37 @@ private void DisposeImage()
...
-                var svg = _img as SvgFragment;
-
-                if (bmp != null)
-                {
-                    return RectangleF.Create(Location.ToDeviceValue(null, this), SizeF.Create(bmp.Width, bmp.Height));
-                }
-                if (svg != null)
+            if (this.ClipPath != null)
+            {
+                SvgClipPath clipPath = this.OwnerDocument.GetElementById<SvgClipPath>(this.ClipPath.ToString());
```
Comment (gentledepp) [inline Svg/Basic Shapes/SvgImage.cs]: can you maybe use `this.ClipPath.Uri.OriginalString` instead? That way we might save some allocation (ToString always allocates a new string)

File: Svg/Basic Shapes/SvgVisualElementExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+namespace Svg.Basic_Shapes
+{
+    public static class SvgVisualElementExtensions
+    {
+        public static RectangleF GetBoundsWithClipPaths(this SvgVisualElement element)
+        {
+            var bounds = element.Path(null).GetBounds();
+            if (element.ClipPath != null)
+            {
+                SvgClipPath clipPath = element.OwnerDocument.GetElementById<SvgClipPath>(element.ClipPath.ToString());
```
Comment (gentledepp) [inline Svg/Basic Shapes/SvgVisualElementExtensions.cs]: another allocation that you may be able to save (`ToString()`)

File: Svg/Paths/SvgPath.cs [hunk 7]
```diff
@@ -115,6 +115,14 @@ protected override bool RequiresSmoothRendering
 
         public override RectangleF GetBounds()
         {
+            if (this.ClipPath != null)
+            {
+                SvgClipPath clipPath = this.OwnerDocument.GetElementById<SvgClipPath>(this.ClipPath.ToString());
```
Comment (gentledepp) [inline Svg/Paths/SvgPath.cs]: another allocation you could save (`ToString()`)

File: Svg/Clipping and Masking/SvgClipPath.cs [hunk 8]
```diff
@@ -73,6 +77,32 @@ public Region GetClipRegion(SvgVisualElement owner)
             return new Region(GetClipRegionPath(owner));
         }
 
+        private RectangleF CalculateClipBounds()
```
Comment (gentledepp) [inline Svg/Clipping and Masking/SvgClipPath.cs:80]: why is the reactangle implementing "GetBounds" differently?

**Review by gentledepp — COMMENTED** [review]
Please fix the commented tiny issues.

Important: Please answer why the rectangle calculates its bounds differently than the rest

File: Svg/Clipping and Masking/SvgClipPath.cs (hunk 8 above)
Comment (Zeljko-Predjeskovic) [reply Svg/Clipping and Masking/SvgClipPath.cs:80]: CliPath is not a VisualElement. I am calculating the bounds based on its children, Like the "svg" element does

Comment (gentledepp) [reply Svg/Clipping and Masking/SvgClipPath.cs:80]: Sry... I was confused because of "Rectangle"... my bad

**Pull request merged**